feat: desktop,extension google login

sidmorizon opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added OAuth login support for desktop applications via local server.
  * Introduced PKCE-based OAuth flows for enhanced security across web, desktop, and extension platforms.
  * Added support for alternative OAuth methods in desktop and extension environments.

* **Improvements**
  * Enhanced session persistence across OAuth flows.
  * Improved OAuth configuration with standardized settings for consistent behavior.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

sidmorizon committed:
feat: implement OneKey ID authentication service and integrate OAuth methods

- Added ServiceOneKeyID for managing OneKey ID authentication, including login with access token, logout, and checking login status.
- Enhanced useSupabaseAuth to support OAuth sign-in methods for Google and Apple.
- Introduced OneKeyIDGallery component for testing OneKey ID authentication features.
- Updated BackgroundApi and BackgroundApiProxy to include the new OneKey ID service.
- Added new routes and components to the Developer Gallery for showcasing OneKey ID functionality.

sidmorizon committed:
feat: enhance OAuth handling in useSupabaseAuth for desktop

- Introduced support for two OAuth methods in the desktop environment: WEBVIEW and DEEP_LINK.
- Updated redirect logic to accommodate both methods, improving user experience and reliability.
- Added detailed comments and documentation for each method's pros and cons.
- Enhanced the UI for the OAuth webview, including a close button and error handling for navigation events.
- Implemented timeout handling for OAuth sign-in processes to improve robustness.

sidmorizon committed:
feat: implement OAuth callback handling and enhance extension support

- Added a new HTML file for OAuth callback handling in Chrome extensions.
- Updated the manifest files to include necessary permissions and OAuth2 configuration.
- Enhanced the OAuth handling logic in the extension to support identity permissions.
- Improved the overall structure and organization of OAuth-related functions across platforms.
- Updated the webpack configuration to include the new OAuth callback page.
- Refactored the useSupabaseAuth hook to streamline OAuth processes for different platforms.

sidmorizon committed:
feat: enhance OAuth session handling and refactor related components

- Introduced a temporary Supabase client to manage OAuth sessions without automatic persistence.
- Updated various OAuth functions to accept options for session persistence, improving flexibility.
- Refactored the `useSupabaseAuth` hook to streamline OAuth processes across platforms.
- Added detailed logging for OAuth session handling to aid in debugging.
- Enhanced the UI components to include session persistence options for better user control.

sidmorizon committed:
feat: enhance OAuth popup handling and session management

- Added a polling mechanism to keep the OAuth popup window focused during authentication, improving user experience.
- Refactored the `openOAuthPopupExtIdentity` function to manage window focus and cleanup more effectively.
- Updated `openOAuthPopupWeb` to allow optional session persistence in the OAuth session parameters.
- Enhanced `useSupabaseAuth` to support optional session persistence for better flexibility.
- Introduced JWT decoding utility in `stringUtils` for improved access token handling in the UI components.

sidmorizon committed:
chore: update .gitignore and remove debug log file

- Added .cursor/debug.log to .gitignore to prevent tracking of debug logs.
- Deleted the .cursor/debug.log file to clean up unnecessary debug information.

sidmorizon committed:
feat: implement local OAuth server for desktop authentication

- Introduced a new local HTTP server for handling OAuth callbacks in desktop applications, enhancing security and user experience.
- Updated the OAuth flow to utilize Google ID tokens and PKCE for secure authentication without requiring a client secret.
- Refactored the `useSupabaseAuth` hook to support the new localhost method, bypassing Supabase redirect URL restrictions.
- Enhanced the `openOAuthPopupDesktopLocalhost` function to manage OAuth sessions effectively, including session persistence options.
- Added IPC event handling for OAuth callbacks to streamline communication between the main and renderer processes.
- Updated documentation and comments to reflect the new authentication flow and its benefits.

sidmorizon committed:
feat: add desktop OAuth lab and refactor authentication flow

- Introduced a new desktop OAuth lab for testing Google OAuth with a local server setup.
- Updated the OAuth flow to utilize a fixed port range for callbacks, enhancing reliability.
- Refactored existing OAuth handling functions to support the new local server method.
- Consolidated OAuth constants into a shared file for better management and consistency.
- Removed deprecated OAuth constants and updated related components to use the new structure.
- Enhanced session management and error handling across various OAuth methods.

sidmorizon committed:
feat: update extension manifest to include development-only extension ID

- Added a new `devExtensionIdKey` for local development, ensuring a stable extension ID for unpacked extensions.
- Updated the manifest to conditionally use the development ID based on the environment, improving OAuth redirect handling for local setups.
- Enhanced comments for clarity on the purpose of the `key` field in the manifest.

sidmorizon committed:
fix: update development extension ID handling in manifest

- Simplified the assignment of the `devExtensionIdKey` in the manifest by removing the conditional check for the production environment.
- Ensured that the development extension ID is always set, improving consistency for local development setups.

sidmorizon committed:
fix: update OAuth callback ports for local development

- Changed the fixed port range for OAuth callbacks to a new set of ports, improving compatibility with local development setups.
- Updated comments to reflect the new port configuration for clarity and ease of use.

sidmorizon committed:
refactor: migrate OAuth server implementation to a dedicated service

- Moved the OAuth server logic from `oauth-server.ts` to a new service located in `service/oauthLocalServer`.
- Introduced new HTML templates for OAuth callback handling, improving the user experience during authentication.
- Updated the main application file to import the new `setMainWindow` function from the refactored service.
- Enhanced the overall structure and organization of OAuth-related functions for better maintainability.

revan-zhang commented:
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scanner | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-critical.svg) Critical | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-high.svg) High | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-medium.svg) Medium | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-low.svg) Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/rayston92/pr-checks/88336a23-73f1-4f3b-ae1b-f4d9925d5f72?refs=cf7cb1221a650ee1406d06d59f75afc091c7b763&source=prChecksComment) |
| :white_check_mark: | Licenses| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/rayston92/pr-checks/88336a23-73f1-4f3b-ae1b-f4d9925d5f72/license?refs=cf7cb1221a650ee1406d06d59f75afc091c7b763&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

sidmorizon committed:
chore: remove desktop OAuth lab files

- Deleted the `.gitignore`, `package.json`, `README.md`, and main source file for the desktop OAuth lab, cleaning up unused components.
- This removal is part of a broader effort to streamline the project and eliminate deprecated features.

sidmorizon committed:
refactor: update OAuth flow to use PKCE for enhanced security

- Refactored the OAuth callback handling to utilize the PKCE flow, replacing token extraction from URL hash with authorization code retrieval from the URL query.
- Updated related components to send and process authorization codes instead of tokens, improving security by not exposing tokens in the URL.
- Enhanced the Supabase client configuration to support PKCE flow, ensuring better session management and security practices.
- Cleaned up and updated comments across the codebase to reflect the new authentication flow and its benefits.

sidmorizon committed:
refactor: make persistSession optional in OAuth functions

- Updated the `persistSession` parameter in multiple OAuth functions to be optional, enhancing flexibility in session management.
- Adjusted related components to accommodate the change, ensuring backward compatibility while allowing for more dynamic session handling.
- Improved code clarity and consistency across the OAuth handling functions.

sidmorizon committed:
refactor: enhance OAuth flow to include state validation for PKCE

- Updated the OAuth callback handling to include state validation, improving security against CSRF attacks.
- Modified components to extract and validate the `state` parameter alongside the `code` during the OAuth process.
- Enhanced error handling to ensure proper responses for state mismatches and missing states.
- Updated comments and documentation to reflect the changes in the OAuth flow and its security implications.

sidmorizon committed:
fix: lint

sidmorizon committed:
refactor: update Supabase project URL and API key based on environment

- Introduced a constant to determine the development environment.
- Updated the Supabase project URL and public API key to switch based on the environment, enhancing flexibility for local testing and production deployment.

sidmorizon committed:
Merge branch 'x' into fix/keyless-google-login

originalix commented on apps/desktop/app/app.ts:
```diff
@@ -562,6 +563,9 @@ async function createMainWindow() {
   void browserWindow.loadURL(src);
 
+  // Set main window reference for OAuth server
+  setMainWindow(browserWindow);
```
这个方法名可以优化下，容易混淆

sidmorizon commented:
@coderabbitai resolve
